command/certificate/needsRenewal.go

```diff
@@ -0,0 +1,135 @@
+package certificate
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"github.com/smallstep/cli/errs"
+	"github.com/urfave/cli"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+func needsRenewalCommand() cli.Command {
+	return cli.Command{
+		Name:      "needs-renewal",
+		Action:    cli.ActionFunc(needsRenewalAction),
+		Usage:     `Check if a certificate needs to be renewed`,
+		UsageText: `**step certificate needs-renewal** <crt_file> or <host_name> [**--expires-in <duration>]`,
+		Description: `**step certificate needs-renewal** Checks certificate expiration from file or from a host if 
+		the certificate is over 66% of its lifetime. If the certificate needs renewal this command will return '0'.
+		If the certificate is not far enough into its life, it will return '1'. If validation fails, or if an error occurs, 
+		this command will produce a non-zero return value.
+## POSITIONAL ARGUMENTS
+
+<crt_file>
+: The path to a certificate to validate.
+
+<host_name>
+: Address of remote host 
+
+## EXIT CODES
+
+This command returns 0 if needing renewal, or returns 1 if the certificate does not need renewal. It will return 255 if any errors occurred
+
+## EXAMPLES
+Check certificate for renewal using custom directory 
+'''
+$ step certificate needs-renewal ./certificate.crt 
+'''
+Check if certificate will expire within a given time 
+'''
+$ step certificate needs-renewal ./certificate.crt --expires-in 1h15m
+'''
+`,
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  "expires-in",
+				Usage: `Check if the certificate expires in given time duration`,
+			},
+		},
+	}
+}
+
+func needsRenewalAction(ctx *cli.Context) error {
+	if err := errs.NumberOfArguments(ctx, 1); err != nil {
+		return err
+	}
+	var (
+		crtFile    = ctx.Args().Get(0)
+		expiresIn  = ctx.String("expires-in")
+		roots      = ctx.String("roots")
+		serverName = ctx.String("servername")
+		cert       *x509.Certificate
+	)
+
+	if addr, isURL, err := trimURL(crtFile); err != nil {
+		return err
+	} else if isURL {
+		peerCertificates, err := getPeerCertificates(addr, serverName, roots, false)
+		if err != nil {
+			os.Exit(255)
```

Comment:
Returning `255` is technically correct. But we want a bit more than that. What we really want is to return an error type that has an attribute where we can store what the return code should be. And then we need to be able to parse that error type wherever this error is handled.

Think of yourself as the user of the program. You run it and all you see is exit code `255`. Well, a bunch of things probably cause that error code, so how do you figure out what went wrong. That's why we return error types. So there's some context for the user as to what went wrong.

Unfortunately, I'm not sure if we have an error type that you could use for this. We may have to write a new one. @mariano do we return non zero or 1 exit status from anywhere else? If not, should we create a new error type and then check for it in main.go?

Reply:
We usually just return a wrapped error in this cases. Commands returning an error will print the error and exit with 1. And we can get the trace using `STEPDEBUG=1`

Reply:
We were going for this functionality I believe:

- return 0 if it's a certificate that needs to be renewed
- return 0 if the certificate is expired
- return 1 if it's a certificate that is valid (timewise) but doesn't need to be renewed
- return 255 for any error

Which is why we needed a different error type that would give 255 as an exit code
os.Exit(255) was there in the mean time to make sure the logic worked right until we knew if there was an error type built like that. 
For cases where exit code 1 is needed should we continue to use os.Exit(1) or should we change it to some kind of error type which doesn't seem right if its not actually an error?

Reply:
See https://github.com/smallstep/cli/pull/495#issuecomment-870048089, I think we should be consistent and keep `1` as exit code for an error. 

In bash scripts is very easy to differentiate exit errors. Unless systemd is an issue we should keep `1` as an error and use `2` for example to say that the certificate is valid.